system/src/commonMain/kotlin/com/splendo/kaluga/system/network/state/NetworkStateRepo.kt

```diff
@@ -0,0 +1,123 @@
+/*
+ Copyright 2020 Splendo Consulting B.V. The Netherlands
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+ */
+
+package com.splendo.kaluga.system.network.state
+
+import co.touchlab.stately.concurrency.AtomicReference
+import com.splendo.kaluga.state.ColdStateRepo
+import com.splendo.kaluga.system.network.BaseNetworkManager
+import com.splendo.kaluga.system.network.Network
+import com.splendo.kaluga.system.network.unknownNetworkOf
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.launch
+
+class NetworkStateRepo(
+    private val networkManagerBuilder: BaseNetworkManager.Builder,
+) : ColdStateRepo<NetworkState>() {
+
+    interface Builder {
+        fun create(): NetworkStateRepo
+    }
+
+    private var _lastKnownNetwork = AtomicReference<Network>(Network.Unknown.WithoutLastNetwork(Network.Unknown.Reason.NOT_CLEAR))
+    internal var lastKnownNetwork: Network
+        get() = _lastKnownNetwork.get()
+        set(value) { _lastKnownNetwork.set(value) }
+
+    internal var networkManager: BaseNetworkManager? = null
+        set(value) {
+            field?.let {
+                if(value == null) {
+                    it.dispose()
+                }
+            }
+            field = value
+        }
+
+    override suspend fun deinitialize(state: NetworkState) {
+        lastKnownNetwork = state.networkType
+        networkManager = null
+    }
+
+    override suspend fun initialValue(): NetworkState {
+        networkManager = networkManagerBuilder.create(::onNetworkStateChange)
+        return when(lastKnownNetwork) {
```

Comment:
Last known network unknown 😆

Reply:
I know, it's funny ahahah...
This is something I also would like to discuss. In Android an object called `Network` from `android.net` already exists (Idk in iOS), this may cause confusion. Should we rename classes that name contains `Network` into something else? @thoutbeckers @Daeda88 @dasdranagon

Reply:
ba0105d